### DISCUSSION TITLE
How to use FromRequest with State

### DISCUSSION THREAD

**Discussion by tsxiaofang** [tokio-rs/axum#1600] [Q&A] [answered]
async fn handler(parts: Multipart, State(s): State<Arc<i32>>) -> impl IntoResponse {

}

let route = Router::new()
        .route("/hello", post(handler).get(handler))
        .with_state(Arc::new(1))；
--------------------------------------------------------------------------------------------
errors like
route("/hello", post(handler).get(handler))
    |                                        --- ^^^^^^^ the trait `axum::handler::Handler<_, _, _>` is not implemented for fn item `fn(axum::extract::Multipart, axum::extract::State<std::sync::Arc<i32>>) -> impl futures::Future<Output = impl axum::response::IntoResponse> {pub_handler::handler}`
    |                                        |
    |                                        required by a bound introduced by this call

**Comment by tsxiaofang** [comment]
Implemented FromRequest object cannot be used with State

**Comment by jplatte** [comment]
Try https://docs.rs/axum/latest/axum/attr.debug_handler.html.

**Comment by davidpdrsn** [comment, marked as answer]
`Multipart` needs to be the last argument:

```rust
async fn handler(State(s): State<Arc<...>>, parts: Multipart) -> impl IntoResponse {}
```

Documented [here](https://docs.rs/axum/latest/axum/extract/index.html#the-order-of-extractors)

**Reply by tsxiaofang** [reply]
ok, State must at first

**Reply by davidpdrsn** [reply]
Well no. _State_ can be the first, second, third, whatever, or even last.

But `Multipart` _must_ be the last.
